src/backend/distributed/shared_library_init.c

```diff
@@ -401,6 +402,20 @@ RegisterCitusConfigVariables(void)
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.enforce_foreign_key_restrictions",
+		gettext_noop("Enforce restrictions while querying distributed/reference "
+					 "tables with foreign keys"),
+		gettext_noop("When enabled, cascading modifications from reference tables "
+					 "to distributed tables are traced and acted accordingly "
+					 "to avoid creating distributed deadlocks and ensure correctness."),
```

Comment:
I'm not sure in what situation we would not want to do this. We can easily violate visibility rules if this is disabled.

Reply:
I had two main motivations:

1) Foreign keys without `CASCADE` option
2) As you've noted below, some of the operations do not really affect each other and if someone needs to disable the checks, we'd have at least a way to provide that:
> For example, a CREATE INDEX on a reference table shouldn't really affect SELECTs or CREATE INDEX on a distributed table in the same transaction, but currently it would be classified as DDL and therefore everything gets serialised.

Reply:
Though I can make this `SHOW_NO_ALL`, would that work?

Reply:
> Though I can make this SHOW_NO_ALL, would that work?

yes, that would work